x/qgb/orchestrator/parser.go

```diff
@@ -0,0 +1,67 @@
+package orchestrator
+
+import (
+	"fmt"
+
+	qgbtypes "github.com/celestiaorg/celestia-app/x/qgb/types"
+	sdkcodec "github.com/cosmos/cosmos-sdk/codec"
+	sdkcodectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/gogo/protobuf/proto"
+	coretypes "github.com/tendermint/tendermint/types"
+)
+
+var _ QGBParserI = QGBParser{}
+
+// FIXME remove this interface after writing tests and see if it is needed or not
+type QGBParserI interface {
+	ParseDataCommitmentConfirm(msg sdktypes.Msg) (qgbtypes.MsgDataCommitmentConfirm, error)
+	ParseValsetConfirm(msg sdktypes.Msg) (qgbtypes.MsgValsetConfirm, error)
+	ParseCoreTx(tx coretypes.Tx) (sdktx.Tx, error)
+	ParseSdkMsg(any *sdkcodectypes.Any) (sdktypes.Msg, error)
+}
+
+type QGBParser struct {
+	codec sdkcodec.Codec
+}
+
+func NewQGBParser(codec sdkcodec.Codec) *QGBParser {
+	return &QGBParser{codec: codec}
+}
+
+func (parser QGBParser) ParseValsetConfirm(msg sdktypes.Msg) (qgbtypes.MsgValsetConfirm, error) {
+	vs, ok := msg.(*qgbtypes.MsgValsetConfirm)
+	if !ok {
+		return qgbtypes.MsgValsetConfirm{},
+			fmt.Errorf("unable to cast sdk message to msg valset confirm")
+	}
+	return *vs, nil
+}
+
+func (parser QGBParser) ParseDataCommitmentConfirm(msg sdktypes.Msg) (qgbtypes.MsgDataCommitmentConfirm, error) {
+	dcc, ok := msg.(*qgbtypes.MsgDataCommitmentConfirm)
+	if !ok {
+		return qgbtypes.MsgDataCommitmentConfirm{},
+			fmt.Errorf("unable to cast sdk message to msg data commitment confirm")
+	}
+	return *dcc, nil
+}
+
+func (parser QGBParser) ParseCoreTx(tx coretypes.Tx) (sdktx.Tx, error) {
+	var sdkTx sdktx.Tx
+	err := proto.Unmarshal(tx, &sdkTx)
+	if err != nil {
+		return sdktx.Tx{}, fmt.Errorf("error while unmarshalling core transaction: %s", err)
+	}
+	return sdkTx, nil
+}
```

Comment:
[not blocking][for future reference]
I'm fine with this, as it doesn't matter, but fwiw this is non-standard and not what we do in the rest of this repo

typically, we refer to the interface 

https://github.com/cosmos/cosmos-sdk/blob/c9b446029fc5e73560a3be966f06453ab33e3e83/types/tx_msg.go#L42-L50

Reply:
If I understand right, that is unmarshalling a tendermint transaction to an sdk one. How do you propose to go from one to another?
The other way I think of it is getting the hash from the tendermint transaction and querying it from the app, but I wanted to avoid that double download of transactions (was done is juno block explorer and was so slow to ingest).
If there is a better way to do, I am really open for it

Reply:
we don't have to use, as I don't see specific benefits other than following the docs, but the standard way that is used by most cosmos-sdk chains is to use the encoding config
```go
func decode(b coretypes.Block) error {
	encCfg := encoding.MakeConfig(app.ModuleEncodingRegisters...)

	for _, rawTx := range b.Data.Txs {
		sdkTx, err := encCfg.TxConfig.TxDecoder()(tx)
		if err != nil {
			return err
		}
		for _, msg := range sdkTx.GetMsgs() {
			url := sdk.MsgTypeURL(msg)
			if !qgbtypes.IsConfirmURL(url) {
				continue
			}
			switch msg.(type) {
			case *qgbtypes.MsgValsetConfirm:

			}
		}
	}
}
```

Reply:
the encoding config is threadsafe iirc

Reply:
That I didn't know is possible, I will update it to this and see if it still works as expected :+1: 
Thanks a lot for your help

Reply:
checking for protobuf string url is also faster than type checking, so it acts as a quick check before we do the heavy type check

Reply:
thanks, the main reason we should imo is there is quite a bit of other logic going on in the [default decoder](https://github.com/cosmos/cosmos-sdk/blob/c9b446029fc5e73560a3be966f06453ab33e3e83/x/auth/tx/decoder.go#L15-L76). Just using protobuf _should_ work, but sticking to what the sdk useses will likely save us from weird bugs in the future. It also should work with not just protubuf encoded txs (so people can sign things with ledgers (not that anyone should be doing that with an orchestrator lolol))

Reply:
it's way better :+1: , I will add the changes shortly